lib/improver/spotdata/site_data.py

```diff
@@ -0,0 +1,246 @@
+        file_path : string
+            Path to json file containing SpotData site information.
+
+        Returns:
+        --------
+        sites : dict
+            Dictionary of site data.
+
+        """
+        data_file = open(file_path, 'r')
+        site_data = json.load(data_file)
+        data_file.close()
+
+        latitude_entries = [i_site for (i_site, site) in enumerate(site_data)
+                            if 'latitude' in site.keys()]
+        longitude_entries = [i_site for (i_site, site) in enumerate(site_data)
+                             if 'longitude' in site.keys()]
+
+        if not latitude_entries or not longitude_entries:
+            raise KeyError('longitude and latitude must be defined for '
+                           'site in site_data file')
+
+        # Only proceed with sites for which both latitude and longitude are
+        # defined.
+        if latitude_entries != longitude_entries:
+            valid_entries = set(latitude_entries).intersection(
+                longitude_entries)
+        else:
+            valid_entries = latitude_entries
+
+        site_data = [site_data[i_site] for i_site in valid_entries]
+
+        self.latitudes = np.array([site['latitude'] for site in site_data])
+        self.longitudes = np.array([site['longitude'] for site in site_data])
+
+        # If altitudes are unset use fill value to indicate they are on the
+        # model orography surface.
+        if 'altitude' in site_data[0].keys():
+            self.altitudes = np.array([site['altitude'] for site in site_data])
+        else:
+            self.altitudes = np.full(len(site_data), -9999)
+
+        # If no site ids are provided, use index of read order.
+        if 'bestdata_id' in site_data[0].keys():
+            self.site_ids = np.array([site['bestdata_id']
+                                      for site in site_data])
+        else:
+            self.site_ids = np.arange(len(site_data))
+
+        utckey = None
+        if 'gmtoffset' in site_data[0].keys():
+            utckey = 'gmtoffset'
+        elif 'utcoffset' in site_data[0].keys():
+            utckey = 'utcoffset'
+        elif 'utc_offset' in site_data[0].keys():
+            utckey = 'utc_offset'
+
+        # Use available UTC offset data, or set data with crude longitude
+        # derived value if it is missing.
+        if utckey is None:
+            self.utc_offsets = set_utc_offset(self.longitudes)
+        else:
+            self.utc_offsets = np.array([site[utckey] for site in site_data])
+
+        return self.construct_site_dictionary()
+
+    def runtime_list(self, latitudes, longitudes,
+                     altitudes=None, site_ids=None):
+        """
+        Use data provided on the command line/controlling suite at runtime.
+
+        Args:
+        -----
+        latitudes/longitudes : lists of floats
+            Lists of coordinates at which to extract data, these define the
+            positions of the SpotData sites.
+
```

Comment:
Maybe. There is lot's of checking to do with this input, and I don't really want to put it in the CLI. There is the potential for variation in input files to contend with and the optional nature of properties like altitude and bestdata_id. As such, I think it is easier to treat the runtime_lists and json file inputs separately. I've consolidated the lists into a dictionary in the CLI to shrink the interface.